hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java

```diff
@@ -51,15 +58,38 @@ public static boolean includeInBaseFetchGroup(
 		if ( value instanceof ToOne ) {
 			final ToOne toOne = (ToOne) value;
 			if ( toOne.isLazy() ) {
-				if ( toOne.isUnwrapProxy() ) {
-					if ( toOne instanceof OneToOne ) {
-						return false;
+				// include it in the base fetch group so long as the config allows
+				// using the FK to create an "enhancement proxy"
+
+				// however a few special cases to handle...
+
+				if ( ! toOne.isReferenceToPrimaryKey() ) {
+					// we do not have a reference to the associated primary-key
+					return false;
+				}
+
+				if ( inheritanceChecker.hasSubclasses( toOne.getReferencedEntityName() ) ) {
+					// the associated type has subclasses - we cannot use the enhanced proxy and will generate a HibernateProxy
+					if ( ! toOne.isUnwrapProxyImplicit() ) {
+						// NO_PROXY was explicitly requested
+						BytecodeLogger.LOGGER.debugf(
+								"`%s#%s` was mapped with LAZY with explicit NO_PROXY but the associated entity (`%s`) has subclasses",
+								bootMapping.getPersistentClass().getEntityName(),
+								bootMapping.getName(),
+								toOne.getReferencedEntityName()
+						);
 					}
-					// include it in the base fetch group so long as the config allows
-					// using the FK to create an "enhancement proxy"
+					// however, select the fk to create the proxy
```

Comment:
In `DefaultLoadEventListener#proxyOrLoad` there is a check for the existence of a `ProxyFactory` [here](https://github.com/hibernate/hibernate-orm/blob/master/hibernate-core/src/main/java/org/hibernate/event/internal/DefaultLoadEventListener.java#L283). 

Do we need to check for a `ProxyFactory` here and factor that into the decision for the return value?

IIUC, this will apply regardless of how `allowEnhancementAsProxy` is set. I think that's fine for master.

When this gets backported, I don't think this should return `true` when `allowEnhancementAsProxy == false`. WDYT?

Reply:
That's a good question.  The thought process here was that the check for ProxyFactory was really just needed for entities with subclasses.  That is maybe not exactly correct yet.  This is obviously the main point of change for this improvement and its tricky to get just right - it comes back to what I mentioned earlier about timing and access to info.  This block is exactly why this is still a draft :)

Keep in mind also that this simply defines what columns should be selected as part of the "base fetch group".  I'm not sure I agree with your statement about `allowEnhancementAsProxy == false`.  Still need to play with it some more

Reply:
OK, I see that the existence of a {{ProxyFactory}} isn't known at the time this is called.